### PR TITLE
[integration] Remote Pilot Logger

### PR DESCRIPTION
**Pull request by martynia**
BEGINRELEASENOTES

*WorkloadManagementSystem
CHANGE: Enable remote pilot logging to Tornado 

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+""" :mod: PilotLoggingAgent
+
+    PilotLoggingAgent sends Pilot log files to an SE
+"""
+
+# # imports
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:9]: These `__future__` imports can be removed (also below).

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+from DIRAC import S_OK, S_ERROR
+from DIRAC.ConfigurationSystem.Client.Helpers import CSGlobals
+from DIRAC.Core.Base.AgentModule import AgentModule
+from DIRAC.Core.DISET.RPCClient import RPCClient
+from DIRAC.Core.Security.Locations import getHostCertificateAndKeyLocation, getCAsLocation
+from DIRAC.DataManagementSystem.Client.DataManager import DataManager
+from DIRAC.WorkloadManagementSystem.Service.TornadoPilotLoggingHandler import TornadoPilotLoggingHandler
+import requests
+
+__RCSID__ = "Id$"
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:23]: Remove this

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import json
+
+from DIRAC import S_OK, S_ERROR
+from DIRAC.ConfigurationSystem.Client.Helpers import CSGlobals
+from DIRAC.Core.Base.AgentModule import AgentModule
+from DIRAC.Core.DISET.RPCClient import RPCClient
+from DIRAC.Core.Security.Locations import getHostCertificateAndKeyLocation, getCAsLocation
+from DIRAC.DataManagementSystem.Client.DataManager import DataManager
+from DIRAC.WorkloadManagementSystem.Service.TornadoPilotLoggingHandler import TornadoPilotLoggingHandler
+import requests
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:21]: Group imports together (this at line 13)

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import os
+import json
+
+from DIRAC import S_OK, S_ERROR
+from DIRAC.ConfigurationSystem.Client.Helpers import CSGlobals
+from DIRAC.Core.Base.AgentModule import AgentModule
+from DIRAC.Core.DISET.RPCClient import RPCClient
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:17]: Not used.

File: src/DIRAC/WorkloadManagementSystem/ConfigTemplate.cfg [hunk 5]
```diff
@@ -17,6 +17,14 @@ Services
       Default = authenticated
     }
   }
+  TornadoPilotLogging
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/ConfigTemplate.cfg:20]: There's an old DISET PilotLogging service that I would just remove (it was never used)

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
+""" :mod: PilotLoggingAgent
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:1]: This agent would also need to be added to the ConfigTemplate.cfg

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        get the target SE name from the CS.
+        Obtain log file location from Tornado.
+
+        :param self: self reference
+        """
+
+        # get shifter proxy for uploads (VO-specific shifter from the Defaults CS section)
+        self.shifterName = self.am_getOption("ShifterName", "GridPPLogManager")
+        self.am_setOption("shifterProxy", self.shifterName)
+        self.uploadSE = self.am_getOption("UploadSE", "UKI-LT2-IC-HEP-disk")
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:46]: These defaults I guess they are only temporary.

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+        self.message = self.am_getOption("Message", "PilotLoggingAgent initialised.")
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:48]: What is this for?

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        self.log.info("message = %s" % self.message)
+        self.certAndKeyLocation = getHostCertificateAndKeyLocation()
+        self.casLocation = getCAsLocation()
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:51]: No need to be instance members.

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        try:
+            with requests.post(self.server, data=data, verify=self.casLocation, cert=self.certAndKeyLocation) as res:
+                if res.status_code not in (200, 202):
+                    message = "Could not get metadata from %s: status %s" % (self.server, res.status_code)
+                    self.log.error(message)
+                    return S_ERROR(message)
+                resDict = res.json()
+        except Exception as exc:
+            message = "Call to server %s failed with %s " % (self.server, exc)
+            self.log.error(message)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:67]: `self.log.exception()` could be used.

File: src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        self.log.info("Request initialised.. ")
+
+    auth_sayHello = ["all"]
+
+    def export_sayHello(self):
+        ## Insert your method here, don't forget the return should be serializable
+        ## Returned value may be an S_OK/S_ERROR
+        ## You don't need to serialize in JSON, Tornado will do it
+        self.log.info("Hello...")
+        return S_OK("Hello!")
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py:55]: To be removed, I guess :)

File: src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        TornadoPilotLoggingHandler.log.info("Pilot logging directory:", logPath)
+
+    def initializeRequest(self):
+        """
+        Called for each request.
+
+        :return: None
+        """
+
+        self.log.info("Request initialised.. ")
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py:46]: Not needed.

File: src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        :return: None
+        """
+
+        TornadoPilotLoggingHandler.log.info("Handler initialised ...")
+        cls.meta = {}
+        logPath = os.path.join(os.getcwd(), "pilotlogs")
+        cls.meta["LogPath"] = logPath
+        if not os.path.exists(logPath):
+            os.makedirs(logPath)
+        TornadoPilotLoggingHandler.log.info("Pilot logging directory:", logPath)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py:37]:
```suggestion
        cls.log.info("Pilot logging directory:", logPath)
```

Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py]: Done.

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py (hunk 7 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py:46]: Yes.

File: src/DIRAC/WorkloadManagementSystem/ConfigTemplate.cfg (hunk 5 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/ConfigTemplate.cfg:22]: I haven't removed it yet.

File: src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py (hunk 12 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py:60]: This comment did not align, after I made changes, so I don't remember what it was referring to ;-)

File: src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py (hunk 11 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/TornadoPilotLoggingHandler.py]: Done

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py (hunk 8 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py]: Gone

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py (hunk 4 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py]: Removed unused imports.

File: src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py (hunk 1 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Agent/PilotLoggingAgent.py]: Gone.

**Comment by martynia** [comment]
Outdated. A new pull request opened (#6208) , based on this one.

**Pull request closed without merging**